CUDADataFormats/Common/interface/HeterogeneousSoA.h

```diff
@@ -0,0 +1,247 @@
+#ifndef CUDADataFormatsCommonHeterogeneousSoA_H
+#define CUDADataFormatsCommonHeterogeneousSoA_H
+
+#include "HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/host_unique_ptr.h"
+
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "HeterogeneousCore/CUDAServices/interface/CUDAService.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/copyAsync.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+
+// a heterogeneous unique pointer...
+template<typename T>
+class HeterogeneousSoA {
+public:
+
+  using Product = T;
+
+  HeterogeneousSoA() = default; // make root happy
+  ~HeterogeneousSoA() = default;
+  HeterogeneousSoA(HeterogeneousSoA&&) = default;
+  HeterogeneousSoA& operator=(HeterogeneousSoA&&) = default;
+
+  explicit HeterogeneousSoA(cudautils::device::unique_ptr<T> && p) : dm_ptr(std::move(p)) {}
+  explicit HeterogeneousSoA(cudautils::host::unique_ptr<T> && p) : hm_ptr(std::move(p)) {}
+  explicit HeterogeneousSoA(std::unique_ptr<T> && p) : std_ptr(std::move(p)) {}
+
+
+  auto const * get() const {
+    return dm_ptr ? dm_ptr.get() : (hm_ptr ? hm_ptr.get() : std_ptr.get());
+  }
+
+  auto const & operator*() const {
+    return *get();
+  }
+
+  auto const * operator->() const {
+    return get();
+  }
+
+
+  auto * get() {
+    return dm_ptr ? dm_ptr.get() : (hm_ptr ? hm_ptr.get() : std_ptr.get());
+  }
+
+  auto &  operator*() {
+    return *get();
+  }
+
+  auto * operator->() {
+    return get();
+  }
+
+
+
+  // in reality valid only for GPU version...
+  cudautils::host::unique_ptr<T>
+  toHostAsync(cuda::stream_t<>& stream) const {
+    assert(dm_ptr);
+    edm::Service<CUDAService> cs;
+    auto ret = cs->make_host_unique<T>(stream);
+    cudaCheck(cudaMemcpyAsync(ret.get(), dm_ptr.get(), sizeof(T), cudaMemcpyDefault, stream.id()));
+    return ret;
+  }
+
+
+
+private:
+  // a union wan't do it, a variant will not be more efficienct
+  cudautils::device::unique_ptr<T> dm_ptr; //!
+  cudautils::host::unique_ptr<T> hm_ptr; //!
+  std::unique_ptr<T> std_ptr; //!
```

Comment:
What is the added benefit of including the device pointer here? Is it only about having `toHostAsync()`?

Reply:
the device_ptr is needed for the GPU implementation (following your implementation elsewhere). If not here, where else?

Reply:
Does `CUDAProduct<cudautils::device::unique_ptr<T>>` work?

Reply:
> Does `CUDAProduct<cudautils::device::unique_ptr<T>>` work?

It had a problem with underscore in the friendly branch name, right?

Ok, fine (for now).

Reply:
interesting alternative (should bypass Root issue), not sure about ```friendly branch name```